File: 1642-Furthest Building You Can Reach.py
import heapq
class Solution(object):
    def furthestBuilding(self, heights, bricks, ladders):
        """
        :type heights: List[int]
        :type bricks: int
        :type ladders: int
        :rtype: int
        """
        minheap = []
        ladderUsed = 0
        for i in range(1, len(heights)):
            diff = heights[i] - heights[i-1]
            if diff <= 0:
                continue
            if ladderUsed < ladders:
                heapq.heappush(minheap, diff)
                ladderUsed += 1
            else:
                heapq.heappush(minheap, diff)
                curr = heapq.heappop(minheap)
                if bricks >= curr:
                    bricks -= curr
                else:
                    return i - 1
        
        return len(heights) - 1

File: C++/1642-Furthest Building You Can Reach.cpp
class Solution {
public:
    int furthestBuilding(vector<int>& heights, int bricks, int ladders) {
        priority_queue<int, vector<int>, greater<int>> pq;
        int N = heights.size();
        int remain = 0;
        
        for (int i=1; i<N; i++) {
            if (heights[i]-heights[i-1] > 0) {
                pq.push(heights[i]-heights[i-1]);
                if (pq.size() > ladders) {
                    remain += pq.top();
                    pq.pop();
                    if (remain > bricks) return i-1;
                }
            }
        }
        return N-1;
    }
};